matcher.go

```diff
@@ -208,3 +209,21 @@ func AllNumberGroupsAreExactlyPresent(
 		strings.HasSuffix(candidateGroups[candidateNumberGroupIndex],
 			formattedNumberGroups[0]))
 }
+
+// Returns whether the given national number (a string containing only decimal digits) matches
+// the national number pattern defined in the given PhoneNumberDesc message.
+func MatchNationalNumber(number string, numberDesc PhoneNumberDesc, allowPrefixMatch bool) bool {
+	nationalNumberPattern := numberDesc.GetNationalNumberPattern()
+	// We don't want to consider it a prefix match when matching non-empty input against an empty pattern.
+	if len(nationalNumberPattern) == 0 {
+		return false
+	}
+	patP := `^(?:` + nationalNumberPattern + `)$` // Strictly match
+	regex := regexFor(patP)
+	return match(number, regex, allowPrefixMatch)
+}
+
+// TODO: review this matches java version?
```

Comment:
Depends on what syntax those patterns. For example Java `\p{XDigit}` would be `[[:xdigit:]]` in Golang... have you looked at the regexes?

Reply:
yeah i'm not an expert on the differences between java/go regexes so would appreciate input on this. The java code is:
``` java
private static boolean match(CharSequence number, Pattern pattern, boolean allowPrefixMatch) {
    Matcher matcher = pattern.matcher(number);
    if (!matcher.lookingAt()) {
      return false;
    } else {
      return (matcher.matches()) ? true : allowPrefixMatch;
    }
  }
```

AFAICT:
in java: `matcher.lookingAt()`'s equivalent in go is:
```go
	ind := matcher.FindIndex(target)
	return len(ind) > 0 && ind[0] == 0 
```
as seen [here (java)](https://github.com/google/libphonenumber/blob/master/java/libphonenumber/src/com/google/i18n/phonenumbers/PhoneNumberUtil.java#L2963) and [here (go)](https://github.com/nyaruka/phonenumbers/blob/main/phonenumbers.go#L2639-L2640) 
and also [here (java)](https://github.com/google/libphonenumber/blob/master/java/libphonenumber/src/com/google/i18n/phonenumbers/PhoneNumberUtil.java#L2924-L2925) and [here (go)](https://github.com/nyaruka/phonenumbers/blob/main/phonenumbers.go#L2608-L2611)
with the an exception that i found [here (java)](https://github.com/google/libphonenumber/blob/master/java/libphonenumber/src/com/google/i18n/phonenumbers/PhoneNumberUtil.java#L3074) [here (go)](https://github.com/nyaruka/phonenumbers/blob/main/phonenumbers.go#L2788)

With that, then the code should be ported to:
```go
func match(number string, pattern *regexp.Regexp, allowPrefixMatch bool) bool {
	ind := pattern.FindStringIndex(number)
	if len(ind) == 0 || ind[0] != 0 {
		return false
	}
	patP := `^(?:` + pattern.String() + `)$` // Strictly match
	pat := regexFor(patP)
	return pat.MatchString(number) || allowPrefixMatch
}
```

Does that make sense to you? Lmk if i missed something

Reply:
That looks right but have you implemented `regexFor`? It's possible we don't need to translate between Java and Go regexes - it depends what syntax they use.

Reply:
`regexFor` is just reusing this function which uses the regex cache: https://github.com/nyaruka/phonenumbers/blob/main/phonenumbers.go#L613-L620

the regexes used are from the metadata xml file `nationalNumberPattern`, e.g. https://github.com/google/libphonenumber/blob/master/resources/ShortNumberMetadata.xml#L4822-L4854

Reply:
Ah ok I thought it was translating the regex. I'm looking thru this xml and I'm not seeing any Java-specific regex syntax so I think we might be good! It's all just digit sequences and non-capturing groups.